apps/api/src/app/inbox/usecases/session/session.usecase.ts

```diff
@@ -161,4 +210,421 @@ export class Session {
 
     return tierLimitMs / 1000 / 60 / 60;
   }
+
+  async isKeylessExpired(applicationIdentifier: string | undefined) {
+    if (!applicationIdentifier) {
+      return true; // If no identifier is provided, consider it expired
+    }
+
+    const parts = applicationIdentifier.replace(this.KEYLESS_ENVIRONMENT_PREFIX, '').split('_');
+    if (parts.length < 1) {
+      return true; // Invalid format, consider expired
+    }
+
+    const createdDate = parts[0];
+
+    if (!createdDate || createdDate.length < 8) {
+      // Ensure we have at least 4 bytes (8 hex chars)
+      return true; // Invalid timestamp format, consider expired
+    }
+
+    try {
+      const createdDateTimestamp = timestampHexToDate(createdDate);
+      const now = new Date();
+      const diffTimeInHours = differenceInHours(now, createdDateTimestamp);
+
+      if (diffTimeInHours > KEYLESS_RETENTION_TIME_IN_HOURS) {
+        return true;
+      }
+    } catch (error) {
+      this.logger.error({ err: error }, 'Error parsing timestamp');
+
+      // If there's any error parsing the timestamp, consider it expired
+      return true;
+    }
+
+    return false;
+  }
+
+  async processKeyless(): Promise<EnvironmentResponseDto> {
+    if (process.env.NOVU_ENTERPRISE !== 'true') {
+      throw new BadRequestException('Keyless is not supported in community edition');
+    }
+
+    const organization = await this.communityOrganizationRepository.findById(process.env.KEYLESS_ORGANIZATION_ID!);
+
+    if (!organization) {
+      this.logger.error('Keyless Organization not found');
+      throw new InternalServerErrorException('Keyless Organization not found');
+    }
+
+    const user = await this.communityUserRepository.findByEmail(process.env.KEYLESS_USER_EMAIL!);
+
+    if (!user) {
+      throw new InternalServerErrorException('Keyless User not found');
+    }
+
+    const key = `sk_${await this.generateUniqueApiKey.execute()}`;
+    const encryptedApiKey = encryptApiKey(key);
+    const hashedApiKey = createHash('sha256').update(key).digest('hex');
+
+    const encodedDate = dateToTimestampHex(new Date());
+    const identifier = `${this.KEYLESS_ENVIRONMENT_PREFIX}${encodedDate}_${shortId(4)}`;
+    const environment = await this.environmentRepository.create({
```

Comment:
I wonder, to simplify the retention script, should we just add some nullable expireAt field?

Reply:
all of this is not relevant to the retention script, the retention script will live separately and delete by environment created at.